File: Python/Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        
        # O(n^2) -> good readability
        max_length = 0
        pos1, pos2 = 0, 0
        
        # for each element, expend its neighbors on both sides, stop if non-matched neighbors are found
        for index in range(len(s)):
            odd_length = 1
            even_length = 0
            
            # odd
            index_left = index - 1
            index_right = index + 1
            while index_left >= 0 and index_right < len(s):
                if s[index_left] == s[index_right]:
                    odd_length += 2
                    index_left -= 1
                    index_right += 1
                else:
                    break
            
            length = index_right - index_left - 1
            if max_length < length:
                max_length = length
                pos1 = index_left + 1
                pos2 = index_right - 1
            
            # even
            index_left = index
            index_right = index + 1
            while index_left >=0 and index_right < len(s):
                if s[index_left] == s[index_right]:
                    even_length += 2
                    index_left -= 1
                    index_right += 1
                else:
                    break

            length = index_right - index_left - 1
            if max_length < length:
                max_length = length
                pos1 = index_left + 1
                pos2 = index_right - 1
        
        
        return s[pos1 : pos2 + 1]
            
            

File: C++/Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {

        // expand every element -> fetch the maximum expansion
        int max_len = 1, pos = 0;
        for(int i = 0, shift = 1, local_max = 1; i < s.length(); i++){
            // odd
            local_max = 1;
            shift = 1;
            while(i - shift >= 0 && i + shift < s.length() && s[i - shift] == s[i + shift]){
                local_max += 2;
                shift++;
            }
            
            if(max_len < local_max){
                max_len = max_len > local_max? max_len: local_max;
                pos = i - shift + 1;
            }
            
            // even
            if(i + 1 < s.length() && s[i + 1] == s[i]){
                local_max = 2;
                shift = 1;
                while(i - shift >= 0 && i + shift + 1 < s.length() && s[i - shift] == s[i + shift + 1]){
                local_max += 2;
                shift++;
            }

                if(max_len < local_max){
                    max_len = max_len > local_max? max_len: local_max;
                    pos = i - shift + 1;
                }
            }
        }
        
        // return substring
        return s.substr(pos, max_len);
    }
};